k2/csrc/intersect_pruned.cu

```diff
@@ -1068,16 +1078,299 @@ class MultiGraphDenseIntersectPruned {
+        int32_t t = begin_t + i;
+        old_states_offsets_data[i] = tot_states;
+        old_arcs_offsets_data[i] = tot_arcs;
+        if (i < num_t) {
+          tot_states += frames_[t]->arcs.TotSize(1);
+          tot_arcs += frames_[t]->arcs.TotSize(2);
+        }
+      }
+    }
 
-    cur_frame->arcs = SubsampleRagged(cur_frame->arcs,
-                                      renumber_cur_arcs);
+
+    // contains respectively: row_splits1_ptrs, row_ids1_ptrs,
+    // row_splits1_ptrs, row_splits2_ptrs,
+    // old_arcs_ptrs (really type ArcInfo*),
+    // old_states_ptrs (really type StateInfo*).
+    Array1<void*> old_all_ptrs(cpu, num_t * 6);
+
+    Renumbering renumber_states(c_, tot_states),
+        renumber_arcs(c_, tot_arcs);
+    {
+      void                    **all_p = old_all_ptrs.Data();
+      int32_t **old_row_splits1_ptrs_data = (int32_t**)all_p,
+                 **old_row_ids1_ptrs_data = (int32_t**)all_p + num_t,
+          **old_row_splits2_ptrs_data = (int32_t**)all_p + 2 * num_t,
+             **old_row_ids2_ptrs_data = (int32_t**)all_p + 3 * num_t;
+      StateInfo **old_states_ptrs_data = (StateInfo**)all_p + 4 * num_t;
+      ArcInfo **old_arcs_ptrs_data = (ArcInfo**)all_p + 5 * num_t;
+      int32_t *old_states_offsets_data = old_states_offsets.Data(),
+                *old_arcs_offsets_data = old_arcs_offsets.Data();
+
+      for (int32_t t = end_t - 1; t >= begin_t; --t) {
+        int32_t i = t - begin_t;
+        Array1<char> this_states_keep =
+            renumber_states.Keep().Arange(old_states_offsets_data[i],
+                                          old_states_offsets_data[i + 1]),
+            this_arcs_keep =
+            renumber_arcs.Keep().Arange(old_arcs_offsets_data[i],
+                                        old_arcs_offsets_data[i + 1]);
+        FrameInfo *cur_frame = frames_[t].get();
+        PropagateBackward(t, cur_frame, frames_[t+1].get(),
+                          &this_states_keep, &this_arcs_keep);
+
+        old_row_splits1_ptrs_data[i] = cur_frame->arcs.RowSplits(1).Data();
+        old_row_ids1_ptrs_data[i] = cur_frame->arcs.RowIds(1).Data();
+        old_row_splits2_ptrs_data[i] = cur_frame->arcs.RowSplits(2).Data();
+        old_row_ids2_ptrs_data[i] = cur_frame->arcs.RowIds(2).Data();
+        old_arcs_ptrs_data[i] = cur_frame->arcs.values.Data();
+        old_states_ptrs_data[i] = cur_frame->states.values.Data();
+
+        // We can't discard any states on t == begin_t because: if it is not t ==
+        // 0, it would be inconvenient to map the dest-states of arcs on t - 1;
+        // and if it is t == 0, this may remove the start-state, which would make
+        // it more complex to avoid invalid FSAs (e.g. with an end-state but no
+        // start-state, or in which we incorrectly interpret a non-start state as
+        // the start state).
+        if (i == 0)  // t == begin_t
+          this_states_keep = (char)1;  // set all elements of the array
```

Comment:
I have been wondering whether this kind of conversion is needed.

Reply:
The cast is likely not needed but I feel it helps clarity.